[dynamo ] Enabling batch_isend_irecv to compile

konradha opened the pull request:
This is a first stab at enabling compilation for operators involving pointwise communications.

I've taken advantage of the traceable functional collectives paradigm a little here and adapted the function call args to `batch_isend_irecv` to not return any work objects. Rather, the graph now contains tensors that are registered with the work objects associated to coalesced groups.

Additionally, `isend` and `irecv` compile as well, maybe we should take them out of the _functional_collectives here. Happy to adapt that.

This feature has been tested on NCCL, RCCL and Gloo backends.

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @xmfan @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela @mlazos @H-Huang @ezyang @chenyang78

konradha committed:
Feature: Compiled batch_isend_irecv, isend, irecv

konradha committed:
Clean up test

konradha committed:
Small change to UserDefinedVariable call priorities

konradha committed:
Address some comments

konradha commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -4849,6 +4850,263 @@ def test_batched_send_recv(self, op_sizes_per_coalesce, timing_enabled):
+        torch.cuda.synchronize()
+
+        if timing_enabled:
+            time.sleep(1)
```
This follows the nomenclature of a similar function I've adapted.

konradha commented on torch/_dynamo/variables/functions.py:
```diff
@@ -1793,6 +1793,36 @@ def call_function(
+            p2p_ops = kwargs["p2p_op_list"]
+
+            if not isinstance(p2p_ops, variables.ListVariable):
+                raise Exception
```
Thank you, forgot about that one.

anijain2305 commented:
cc @wconstab @kwen2501 for initial review on the distributed code part. I can review the Dynamo codebase (but I need to learn more about the distributed part to review).

ezyang commented:
There was a more minimal PR at https://github.com/pytorch/pytorch/pull/155070 which I think I'd prefer to land first, as this one is a lot more involved

konradha committed:
Linter fixes

ezyang commented:
@xmfan @sanketpurandare We have made progress on pipelining support in the compiler. Can you two take a look here and make a call?

xmfan commented:
I suggest we either gate the automatic dynamo rewrite part of this PR under some flag or remove it alltogether.

I'm good with adding `batch_isend_irecv`, `isend` and `irecv` as funcols. If the user manually writes them using funcols, the compiler should be able to capture them properly, so the Dynamo variable changes are good too.

However, I believe more verification is needed to land the Dynamo's automatic rewrite of collective. I'm worried that the compiler will attempt to reorder them and existing distributed code that relied on Dynamo graph breaks for p2p ends up hanging. Even when we roll it out, we would want a config to turn off.

konradha committed:
Force dependencies for effectful P2P ops; add guard to explicitly enable P2P comms compilation

xmfan commented on torch/distributed/_functional_collectives.py:
```diff
@@ -1188,4 +1314,19 @@ def all_gather_inplace(
+_register_effectful_op(torch.ops._c10d_functional.irecv.default, _EffectType.ORDERED)
+_register_effectful_op(
+    torch.ops._c10d_functional.batch_p2p_ops.default, _EffectType.ORDERED
+)
```
these are not needed if you wrap the output of your ops w/ `_maybe_wrap_tensor` like the other implementations in this file

xmfan commented on torch/distributed/_functional_collectives.py:
```diff
@@ -1005,6 +1056,15 @@ def _reduce_scatter_tensor_coalesced_native_meta(
+
+torch.fx.node.has_side_effect(torch.ops._c10d_functional.isend)
+torch.fx.node.has_side_effect(torch.ops._c10d_functional.irecv)
+torch.fx.node.has_side_effect(torch.ops._c10d_functional.batch_p2p_ops)
```
these are not needed if you wrap the output of your ops w/ `_maybe_wrap_tensor`, when traced, the outputs of your new ops should be used by some `torch.ops._c10d_functional.wait_tensor` op

konradha replied:
If we issue a wait_tensor for every call we remove any opportunity to actually interleave communication and computation, or am I missing something? There will also be some trailing wait/wait_tensor calls then, I think.

The way I'm seeing it is this:

user has something like

```python
works = dist.batch_isend_irecv( <list of P2POp> )
# some work, the batched comms can be issued already
for w in works: w.wait()
```

which then gets transformed to 

```python
work_tensors = dist.batch_p2p_ops( <lists materializing the args to previous P2POp> )
##
for wt in work_tensors: wait_tensor(wt)
```

so the default assumption is that the user's code is correct wrt. liveness and using data that's correctly synchronized or waited on.

xmfan commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -4849,6 +4850,334 @@ def test_batched_send_recv(self, op_sizes_per_coalesce, timing_enabled):
+    )
+    @parametrize("timing_enabled", [True, False])
+    def test_batched_send_recv_compiled(self, op_sizes_per_coalesce, timing_enabled):
+        os.environ["TORCHDYNAMO_ENABLE_P2P_COMPILATION"] = "1"
```
please use the config, and use @torch._dynamo.config.patch. otherwise it leaks into other tests

konradha replied:
Thanks

konradha committed:
Merge branch 'main' into konradha-compileable-async-comms-current

konradha committed:
Address some issues and tests; remove dead code in distributed/_functional_collectives.py

konradha committed:
Add missing file

konradha committed:
Merge branch 'main' into konradha-compileable-async-comms-current

konradha committed:
remove effect tokens, make inductor aware of the new ops

konradha committed:
Update comments

konradha committed:
Merge remote-tracking branch 'origin/main' into konradha-compileable-async-comms-current

konradha committed:
some polish

konradha commented:
@xmfan @wconstab @bdhirsh @ezyang @anijain2305 @sanketpurandare  Summary of where this stands and addressing some comments:

- P2P compilation is turned off by default, requires `TORCHDYNAMO_ENABLE_P2P_COMPILATION=1`
- Effect tokens: per @xmfan's feedback, dropped the explicit `_register_effectful_op` / `has_side_effect`, using `_maybe_wrap_tensor` + `wait_tensor` for ordering, consistent with how all other funcols handle it -- took me some time to get my head around how this works
- `isend`, `irecv`, `batch_isend_irecv` are in `traceable_collective_remaps`. `batch_isend_irecv` gets rewritten inside `CollectiveFunctionRewriteVariable`. The `P2POpVariable` is needed to materialize the stateful `P2POp` into flat args  for the funcol schema since `P2POp` itself can't go into the graph.
- `isend` / `irecv` / `batch_p2p_ops` have proper lowerings. `irecv` uses `create_inplace` with mutation tracking. `isend` creates a `_CollectiveKernel` without mutation since it only reads.
- `batch_p2p_ops` in C++ handles coalescing via `startCoalescing` / `endCoalescing` and registers work on all result tensors.
- #155070 is closed (went stale Dec 2025). This PR covers `isend` / `irecv` / `batch_isend_irecv`; that one only covered blocking `send` / `recv`.

The remaining open item is the isend schema aliasing ([see comment](https://github.com/pytorch/pytorch/pull/161213#discussion_r2465659914)). `isend` returns its input but the schema doesn't declare aliasing. The conservative fix would be `Tensor(a!)`, but that's semantically wrong since isend only reads. Pointwise P2P ops don't map cleanly onto the funcol convention in my view


Here's a benchmark comparing eager, compilation with p2p graph breaks, and with my feature enabled (including some passes in inductor explicitly enabled).
<img width="4800" height="1800" alt="bench_ring_attention" src="https://github.com/user-attachments/assets/687ae785-acfd-453e-8472-174577034a37" />

Corresponding file here, launched with `torchrun --nproc-per-node 4 --standalone` on a 4xH100 node
[bench_ring.py](https://github.com/user-attachments/files/25325655/bench_ring.py)

wconstab commented on torch/csrc/distributed/c10d/Functional.cpp:
```diff
@@ -275,6 +275,99 @@ at::Tensor broadcast(
+  std::vector<at::Tensor> input_wrap = {input};
+  auto work = group->send(input_wrap, dst, tag);
+  c10d::register_work(input, work);
+  return input;
```
well, actually the send op doesn't even return a tensor does it? i think only when we shoehorn it into funcol it wants to return an ASyncTensor that holds its 'work' obj, but this is a bit of a stretch.

wconstab commented:
> The remaining open item is the isend schema aliasing (https://github.com/pytorch/pytorch/pull/161213#discussion_r2465659914). isend returns its input but the schema doesn't declare aliasing. The conservative fix would be Tensor(a!), but that's semantically wrong since isend only reads. Pointwise P2P ops don't map cleanly onto the funcol convention in my view

can you explain this more? isend is basically just reading from its input tensor, isn't it? how is that different from other compute ops, which read their inputs (and do not mutate them)?  What do we do with functional broadcast, for example?  I guess we alloc a new output tensor so there is no aliasing relationship.  For functional send, there is no need to return the self tensor- i guess we could just alloc an empty tensor and wrap that in async collective tensor?

konradha committed:
Merge branch 'main' into konradha-compileable-async-comms-current

konradha committed:
Fix isend aliasing

wconstab reviewed (commented):
i think this looks ok now.  @xmfan @d4l3k @Skylion007 do you want to take a review?

konradha committed:
Make linter happy

konradha committed:
Refactor for nicer linting

konradha committed:
Merge remote-tracking branch 'origin/main' into konradha-compileable-async-comms-current

xmfan commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -6034,6 +6034,404 @@ def test_batched_send_recv(self, op_sizes_per_coalesce, timing_enabled):
+        diff = (out_e - out_c).abs()
+        local_max = diff.max()
+        dist.all_reduce(local_max, op=dist.ReduceOp.MAX)
+        self.assertLess(float(local_max.item()), 1e-5)
```
can we use assertEqual e-3?

konradha committed:
Address some comments

konradha committed:
Return values from remapped ops and align inductor lowerings with empty(0) schema; add graph checks to test.dynamo.test_fake_distributed

xmfan commented on torch/_dynamo/variables/distributed.py:
```diff
@@ -334,3 +335,79 @@ def _setup_hook(
+            hints=[],
+        )
+
+    def var_getattr(self, tx, name):
```
do you mind adding a test where we mutate a P2POp before calling the collective? it should graph break

xmfan commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -804,6 +784,22 @@ def call_method(
         if name == "__eq__" and isinstance(args[0], UserDefinedClassVariable):
             return variables.CONSTANT_VARIABLE_FALSE
 
+        if name == "wait":
```
checking my understanding, is this to avoid graph break when the user writes directly using functional collectives?

xmfan commented on torch/_dynamo/graph_break_registry.json:
where was this graph break fixed?

xmfan commented on test/dynamo/test_fake_distributed.py:
```diff
@@ -158,6 +162,432 @@ def fn(x):
 
+
+@skipIf(not dist.is_available(), "requires distributed")
+class TestFakeDistributedP2P(DynamoTestCase):
```
can we avoid duplicating mostly the same tests between here and test_c10d_nccl.py?

konradha committed:
Address some comments; remove extra variable additions in dynamo

konradha committed:
Nicer tests

konradha committed:
Merge branch 'main' into konradha-compileable-async-comms-current

xmfan reviewed (commented):
We are very close, mostly lgtm, last set of comments

xmfan commented on test/distributed/test_c10d_functional_native.py:
```diff
@@ -257,6 +257,29 @@ def test_functional_collectives_inference_mode(self) -> None:
+        with torch.inference_mode():
+            input = torch.ones((10), device=self.device)
+            output = f(input)
+            self.assertEqual(output, 0 * input)
```
Yeah we can just change the check for the isend rank

konradha committed:
Make test rank-dependent

xmfan reviewed (approved):
Big improvement! Thank you for pulling through and contributing this change

konradha commented:
@xmfang not sure whether a diff like the would fix that -- cannot reproduce this on my machine. Guarding when distributed is not available or not initialized?

```bash
@@ -537,7 +543,7 @@ class UserDefinedClassVariable(UserDefinedVariable):

         constant_args = check_constant_args(args, kwargs)

-        if self.value is torch.distributed.P2POp:
+        if P2POp is not None and self.value is P2POp:
             if not config.enable_p2p_compilation:
                 unimplemented(
                     gb_type="P2P compilation disabled for P2POp construction",
@@ -1399,7 +1405,7 @@ class UserDefinedObjectVariable(UserDefinedVariable):
             "an AttributeMutation mutation_type"
         )

-        if type(self.value) is torch.distributed.P2POp and (
+        if type(self.value) is P2POp and (
             tx.output.side_effects.has_pending_mutation_of_attr(self, name_str)
             or name_str in self.value.__dict__
         ):
```

xmfan commented:
Yeah, mac builds by default don't build with distributed package. We need to gate accesses into torch.distributed.X by first checking torch.distributed.is_available

xmfan commented on torch/_dynamo/variables/user_defined.py:
```diff
@@ -1379,6 +1399,19 @@ def method_setattr_standard(
             "an AttributeMutation mutation_type"
         )
 
+        if type(self.value) is torch.distributed.P2POp and (
```
here
```suggestion
        if torch.distributed.is_available() and type(self.value) is torch.distributed.P2POp and (
```

xmfan committed:
Apply suggestions from code review

Co-authored-by: Simon Fan <simon.xmfan@gmail.com>

konradha committed:
ruff format to make linter happy

xmfan commented:
@pytorchbot merge
